pyvista/core/filters/data_set.py

```diff
@@ -7193,16 +7202,41 @@ def transform(  # type: ignore[misc]
         _update_alg(f, progress_bar, 'Transforming')
         res = pyvista.core.filters._get_output(f)
 
-        # make the previously active scalars active again
-        if active_point_scalars_name is not None:
-            self.point_data.active_scalars_name = active_point_scalars_name
-            res.point_data.active_scalars_name = active_point_scalars_name
-        if active_cell_scalars_name is not None:
-            self.cell_data.active_scalars_name = active_cell_scalars_name
-            res.cell_data.active_scalars_name = active_cell_scalars_name
+        def _restore_active_scalars(input_: ConcreteDataSetType, output_: ConcreteDataSetType):
+            # make the previously active scalars active again
+            input_.point_data.active_scalars_name = active_point_scalars_name
+            input_.cell_data.active_scalars_name = active_cell_scalars_name
+
+            # Only update output if necessary
+            if input_ is not output_:
+                output_.point_data.active_scalars_name = active_point_scalars_name
+                output_.cell_data.active_scalars_name = active_cell_scalars_name
+
+        if isinstance(self, pyvista.RectilinearGrid):
+            output: ConcreteDataSetType = pyvista.StructuredGrid()  # type: ignore[assignment]
```

Comment:
This error should not be ignored. The TypeVar assumes a `T -> T` mapping and therefore expects a `RectilinearGrid`  -> `RectilinearGrid`  mapping, which is not the case. We should use the Union from #6918 instead.
```suggestion
            output: ConcreteDataSetAlias = pyvista.StructuredGrid()
```